### PR TITLE
DRA: migrate DeviceTaint and Conditions to declarative validation

### LINKED ISSUE #135073: [Umbrella] DRA Declarative Validation Migration Tracking Issue
### What would you like to be added?

Issue tracks known DV cases that should tagged for the DRA `resource` group types:

| DRA types.go link | DRA field name | Header Column 3 |
| :--- | :--- | :--- |
| [resource/v1/types.go#L343-L344](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/resource/v1/types.go#L343-L344)| Taints []DeviceTaint  | +k8s:maxLength=16 |
| [resource/v1/types.go#L343-L344](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/resource/v1/types.go#L312-L313)| Conditions []metav1.Condition  | // +k8s:maxItems=8 // +k8s:listType=map // +k8s:listMapKey=type // +k8s:unique=type |

### Why is this needed?

This is needed as part of DV validation migration for more decllarative APIs allowing for additional automated API tooling and improved readability/review-ability

### PR DESCRIPTION
**Pull request by dumko2001**
What type of PR is this?
/kind cleanup
/kind feature

What this PR does / why we need it:
This PR migrates the validation for Device.Taints and AllocatedDeviceStatus.Conditions from manual Go code to declarative markers, addressing items from the tracking issue #135073.

Changes:

staging/src/k8s.io/api/resource/v1/types.go:
Added // +k8s:maxItems=16 to Device.Taints.
Added // +k8s:maxItems=8, // +k8s:listType=map, and // +k8s:listMapKey=type to AllocatedDeviceStatus.Conditions.
pkg/apis/resource/validation/validation.go:
Updated validateDevice to mark Taints size as covered by declarative validation (sizeCovered).
Removed the manual length check for Conditions in validateDeviceStatus.
Ran make update to synchronize generated files (zz_generated.* and generated.proto).

Which issue(s) this PR is related to:
Fixes #135073

Special notes for your reviewer:
The tracking issue suggested adding // +k8s:unique=map for AllocatedDeviceStatus.Conditions.

I have omitted this specific tag because it causes a conflict in update-codegen:

tag "k8s:unique": unique tag may not be used with listType=set or listType=map
Since listType=map already enforces uniqueness on the key (type), the explicit unique tag is redundant.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/resource/validation/validation.go [hunk 1]
```diff
@@ -1282,9 +1282,6 @@ func validateDeviceStatus(device resource.AllocatedDeviceStatus, fldPath *field.
 	if !allocatedDevices.Has(sharedDeviceID) {
 		allErrs = append(allErrs, field.Invalid(fldPath, sharedDeviceID, "must be an allocated device in the claim"))
 	}
-	if len(device.Conditions) > resource.AllocatedDeviceStatusMaxConditions {
-		allErrs = append(allErrs, field.TooMany(fldPath.Child("conditions"), len(device.Conditions), resource.AllocatedDeviceStatusMaxConditions))
```
Comment (lalitc375) [inline pkg/apis/resource/validation/validation.go:1409]: Hand written validation should not be removed. it should be marked by `.MarkCoveredByDeclarative()` Look other references.

File: staging/src/k8s.io/api/resource/v1/types.go [hunk 2]
```diff
@@ -371,6 +371,7 @@ type Device struct {
 	// +optional
 	// +listType=atomic
```
Comment (lalitc375) [inline staging/src/k8s.io/api/resource/v1/types.go:402]: All API versions needs to be updated with these tags i.e. v1beta1 and v1beta2.

File: staging/src/k8s.io/api/resource/v1/types.go [hunk 3]
```diff
@@ -371,6 +371,7 @@ type Device struct {
...
 	// +featureGate=DRADeviceTaints
+	// +k8s:maxItems=16
```
Comment (lalitc375) [inline staging/src/k8s.io/api/resource/v1/types.go]: Add tests for these new validation tags in pkg/registry/resource/resourceslice/declarative_validation_test.go

**Review by lalitc375 — COMMENTED** [review]
Thank you for the PR @dumko2001 . Totally appreciate it.

Please follow this pattern for adding tags.

```
The ideal PR structure for a PR contributing a DV tag migration is:

Commit # 1: IF NOT ALREADY WIRED UP FOR DV - Wire up the apigroup, add empty test and empty generated file. I don't need to review this, you know the pattern by now.

Commit # 2: Use a single DV tag on a single field. E.g. Add +k8s:required to required field. Includes generated code and robust test cases. Emphasis on test-cases -- we want enough to prove that a field is correct, but we don't want to duplicate all the per-tag tests (e.g. a k8s-short-name field has dozens of corner cases, but we really only need a few). We want each case to be as short and specific as possible, which is why the "tweak" pattern and the use of Origin for Invalid() errors is so important. Specify as little as possible to be sufficiently precise. Newcomers need guidance on this.

Commits # 3...N: Use DV tags on the same field, one per commit; repeat until that field is done or all that is left is not handled by DV yet. Includes generated code and tests for each field. (not the case for the specific examples below as they are for single tag migrations)

Commits N+1...M: Repeat the above for more fields. (not the case for the specific examples below as they are for single field migrations)

The emphasis is on lots of small commits. Maintaining commit discipline is hard but worthwhile.

If DV exposes bugs in the existing testing or the hand-written validation itself (!!), those should be fixed as a PR of its own or as commits at the front of the PR.
```

File: pkg/apis/resource/validation/validation.go [hunk 4]
```diff
@@ -1283,9 +1283,16 @@ func validateDeviceStatus(device resource.AllocatedDeviceStatus, fldPath *field.
 		allErrs = append(allErrs, field.Invalid(fldPath, sharedDeviceID, "must be an allocated device in the claim"))
 	}
 	if len(device.Conditions) > resource.AllocatedDeviceStatusMaxConditions {
-		allErrs = append(allErrs, field.TooMany(fldPath.Child("conditions"), len(device.Conditions), resource.AllocatedDeviceStatusMaxConditions))
+		allErrs = append(allErrs, field.TooMany(fldPath.Child("conditions"), len(device.Conditions), resource.AllocatedDeviceStatusMaxConditions).WithOrigin("maxItems").MarkCoveredByDeclarative())
+	}
+	// Allow declarative validation to handle duplicates (via +listType=map)
+	for _, err := range metav1validation.ValidateConditions(device.Conditions, fldPath.Child("conditions")) {
```
Comment (lalitc375) [inline pkg/apis/resource/validation/validation.go]: You should add options in the ValidationConditions method to mark which validations are covered by DV. Look into ValidateCSIDriverName method for example for reference.

https://github.com/kubernetes/kubernetes/blob/b2ac9e206fdd912f35f2ab5b3c5b5243303ba14b/pkg/apis/core/validation/validation.go#L1817

File: staging/src/k8s.io/api/resource/v1/types.go [hunk 5]
```diff
@@ -650,6 +651,7 @@ type DeviceTaint struct {
 	// Must be a label name.
 	//
 	// +required
+	// +k8s:validation:MinLength=1
```
Comment (lalitc375) [inline staging/src/k8s.io/api/resource/v1/types.go]: The tag `+k8s:validation:MinLength` is **not a valid tag** for core API types. The `validation:` namespace is used by controller-gen for CRDs but is not recognized by the Kubernetes core API machinery. Please remove this tag or replace it with a supported one if the machinery allows (e.g., `+k8s:minLength=1` if supported, otherwise keep it handwritten).

File: staging/src/k8s.io/api/resource/v1/types.go [hunk 6]
```diff
@@ -667,6 +669,7 @@ type DeviceTaint struct {
 	//
 	// +required
 	// +k8s:required
+	// +k8s:validation:Enum=NoSchedule;PreferNoSchedule;NoExecute
```
Comment (lalitc375) [inline staging/src/k8s.io/api/resource/v1/types.go]: This tag is invalid (same reason as above). Furthermore, the type `DeviceTaintEffect` already has the `+k8s:enum` tag, which is the standard declarative way to handle enum validation. This field-level override is redundant and should be removed.

**Review by lalitc375 — COMMENTED** [review]
The migration of handwritten validation to declarative validation for DRA is mostly correct for the `Conditions` field. However, there are some issues to address:

1. **Invalid Tags**: The `+k8s:validation:*` tags are not supported in core API types.
2. **Missing +k8s:optional**: `AllocatedDeviceStatus.Conditions` is missing the declarative `+k8s:optional` tag.
3. **Handwritten Coverage**: The `ValidateLabelName` call for `taint.Key` in `validation.go` should remain handwritten and not be marked as covered until a valid declarative tag is implemented.
4. **Test Regression**: Duplicate condition type testing was removed and not re-added.

File: pkg/apis/resource/validation/validation.go (hunk 4 above)
Comment (lalitc375) [inline pkg/apis/resource/validation/validation.go:1416]: You should add options in the ValidationConditions method to mark which validations are covered by DV. Look into ValidateCSIDriverName method for example for reference.

https://github.com/kubernetes/kubernetes/blob/b2ac9e206fdd912f35f2ab5b3c5b5243303ba14b/pkg/apis/core/validation/validation.go#L1817

File: staging/src/k8s.io/api/resource/v1/types.go (hunk 5 above)
Comment (dumko2001) [reply staging/src/k8s.io/api/resource/v1/types.go]: Done. Removed the invalid tag.

File: staging/src/k8s.io/api/resource/v1/types.go (hunk 6 above)
Comment (dumko2001) [reply staging/src/k8s.io/api/resource/v1/types.go]: Done. Removed the redundant tag since 

DeviceTaintEffect
 already handles it declaratively.

File: pkg/apis/resource/validation/validation.go (hunk 4 above)
Comment (dumko2001) [reply pkg/apis/resource/validation/validation.go:1416]: I've reverted the duplicate validation logic to use the standard metav1validation.ValidateConditions directly without marking it as covered by declarative validation.

This ensures duplicate conditions are still correctly caught by the existing handwritten validation logic, avoiding any regression, without requiring invasive changes to the shared metav1 package or complex wrappers at this stage.

File: pkg/apis/resource/validation/validation.go (hunk 1 above)
Comment (dumko2001) [reply pkg/apis/resource/validation/validation.go]: done

File: staging/src/k8s.io/api/resource/v1/types.go (hunk 2 above)
Comment (dumko2001) [reply staging/src/k8s.io/api/resource/v1/types.go]: done

File: staging/src/k8s.io/api/resource/v1/types.go (hunk 3 above)
Comment (dumko2001) [reply staging/src/k8s.io/api/resource/v1/types.go]: done

File: staging/src/k8s.io/api/resource/v1/types.go [hunk 7]
```diff
@@ -1966,6 +1967,10 @@ type AllocatedDeviceStatus struct {
 	// +optional
 	// +listType=map
 	// +listMapKey=type
+	// +k8s:optional
+	// +k8s:listType=map
+	// +k8s:listMapKey=type
```
Comment (lalitc375) [inline staging/src/k8s.io/api/resource/v1/types.go:2144]: I don't think tests are present for 

```
// +k8s:listMapKey=type
```

File: test/declarative_validation/resource/resourceslice/declarative_validation_test.go [hunk 8]
```diff
@@ -91,6 +91,12 @@ func TestDeclarativeValidate(t *testing.T) {
 							resource.DeviceTaintEffect("Invalid"), []string{}),
 					},
 				},
+				"invalid: max taints exceeded": {
```
Comment (lalitc375) [inline test/declarative_validation/resource/resourceslice/declarative_validation_test.go:97]: Also add a valid at limit test case

File: pkg/registry/resource/resourceclaim/declarative_validation_test.go [hunk 9]
```diff
@@ -1738,3 +1739,76 @@ func tweakStatusDevicesTooManyIPs(count int) func(rc *resource.ResourceClaim) {
...
+		old          resource.ResourceClaim
+		expectedErrs field.ErrorList
+	}{
+		"invalid: too many conditions": {
+			update: func() resource.ResourceClaim {
+				claim := mkResourceClaimWithStatus(tweakStatusDevices(resource.AllocatedDeviceStatus{
+					Driver: "test", Pool: "test", Device: "test",
+					Conditions: validConditions,
+				}))
+				claim.ObjectMeta.ResourceVersion = "1"
```
Comment (lalitc375) [inline pkg/registry/resource/resourceclaim/declarative_validation_test.go]: This can move to the test.

File: pkg/registry/resource/resourceclaim/declarative_validation_test.go [hunk 10]
```diff
@@ -1738,3 +1739,76 @@ func tweakStatusDevicesTooManyIPs(count int) func(rc *resource.ResourceClaim) {
...
+			LastTransitionTime: v1.Now(),
+		}
+	}
+
+	testCases := map[string]struct {
+		update       resource.ResourceClaim
+		old          resource.ResourceClaim
+		expectedErrs field.ErrorList
+	}{
+		"invalid: too many conditions": {
```
Comment (lalitc375) [inline pkg/registry/resource/resourceclaim/declarative_validation_test.go]: Also add an atlimit test case.

File: pkg/registry/resource/resourceslice/declarative_validation_test.go [hunk 11]
```diff
@@ -91,6 +91,12 @@ func TestDeclarativeValidate(t *testing.T) {
 							resource.DeviceTaintEffect("Invalid"), []string{}),
 					},
 				},
+				"invalid: max taints exceeded": {
```
Comment (dumko2001) [reply pkg/registry/resource/resourceslice/declarative_validation_test.go]: Done

**Comment by dumko2001** [comment]
@lalitc375 Addressed all 4 comments:

Added +k8s:listMapKey=type test with normalization rule
Added at-limit taints test (16 taints)
Moved ResourceVersion to test loop
Added at-limit conditions test (8 conditions)

File: pkg/apis/resource/validation/validation.go (hunk 4 above)
Comment (lalitc375) [reply pkg/apis/resource/validation/validation.go:1416]: This comment  still need work.

File: pkg/apis/resource/validation/validation.go [hunk 12]
```diff
@@ -64,6 +64,13 @@ var ResourceNormalizationRules = []field.NormalizationRule{
 		Regexp:      regexp.MustCompile(`spec.devices\[(\d+)\]\.basic\.`),
 		Replacement: "spec.devices[$1].",
 	},
+	{
+		// Normalize declarative listMapKey duplicate errors (conditions[N]) to match
+		// handwritten ValidateConditions errors (conditions[N].type)
+		// This is applied to declarative errors (got) to match handwritten format (want)
+		Regexp:      regexp.MustCompile(`\.conditions\[(\d+)\]$`),
+		Replacement: ".conditions[$1].type",
```
Comment (lalitc375) [inline pkg/apis/resource/validation/validation.go:78]: Remove .type from handwritten code. We have decided to not use the key name to show duplicate errors or any errors related to the list constraints.  Fieldpaths in errors can be changed,as they are not considered part of the API.

File: test/declarative_validation/resource/resourceslice/declarative_validation_test.go [hunk 13]
```diff
@@ -91,6 +91,15 @@ func TestDeclarativeValidate(t *testing.T) {
 							resource.DeviceTaintEffect("Invalid"), []string{}),
 					},
 				},
+				"valid: at limit taints": {
```
Comment (lalitc375) [inline test/declarative_validation/resource/resourceslice/declarative_validation_test.go:94]: Update tests are missing for resourceSlice.

File: pkg/registry/resource/resourceclaim/declarative_validation_test.go [hunk 14]
```diff
@@ -1738,3 +1739,120 @@ func tweakStatusDevicesTooManyIPs(count int) func(rc *resource.ResourceClaim) {
...
+
+func TestDeclarativeValidateStatus(t *testing.T) {
+	for _, apiVersion := range apiVersions {
+		t.Run(apiVersion, func(t *testing.T) {
+			testDeclarativeValidateStatus(t, apiVersion)
+		})
+	}
+}
+
+func testDeclarativeValidateStatus(t *testing.T, apiVersion string) {
```
Comment (lalitc375) [inline pkg/registry/resource/resourceclaim/declarative_validation_test.go]: Do we need creation tests for status?

File: pkg/apis/resource/validation/validation.go (hunk 4 above)
Comment (dumko2001) [reply pkg/apis/resource/validation/validation.go]: Done. Duplicate errors marked as covered by DV.

File: pkg/registry/resource/resourceclaim/declarative_validation_test.go (hunk 14 above)
Comment (dumko2001) [reply pkg/registry/resource/resourceclaim/declarative_validation_test.go]: No ,status only has update path, no create operation.

**Comment by lalitc375** [comment]
Please rebase the PR.

**Comment by dumko2001** [comment]
@lalitc375 rebased PR is now green on `a7b6b48f36ea`.

The small follow-up in `test/e2e/dra/kind.yaml` was only to keep the existing DRA kind presubmits working after the rebase onto current master: kinds `kubeadm.k8s.io/v1beta4` `apiServer.extraArgs` needed the new list form for `runtime-config`, otherwise the legacy DRA claim/template APIs used by those jobs were not served and the DRA kind jobs failed with 404s.

Could you please take another look when you have a moment?

**Comment by lalitc375** [comment]
I just realized this has overlap with #139101  . Can you remove DV from conditions in this PR? Please also rebase
